Support SPN attribute for signals

approximatenumber opened the pull request:
Added SPN attribute to signal attributes.
Added SPN support to tests. But currently tests are broken in `master`, only python-tests are passed successfully.

eerimoq commented on cantools/database/can/signal.py:
```diff
@@ -387,6 +389,19 @@ def multiplexer_signal(self):
+    @property
+    def spn(self):
+        """The SPN (Suspect Parameter Number) value if the signal
+        has this attribute, ``None`` otherwise.
```
You should mention J1939.

approximatenumber replied:
You may that I need to validate that DBC file is based on J1939 before SPN parsing?

eerimoq replied:
Nothing like that. Just Mention J1939 in the docstring. SPN is part of the J1939 specification.

approximatenumber replied:
Got it, mentioned.

eerimoq commented:
Just add a test case (and/or extend an existing test case) and I'll merge.

approximatenumber commented:
> Just add a test case (and/or extend an existing test case) and I'll merge.

Added SPN attribute test to existing `test_attributes`
Added SPN attribute to `attributes.dbc`

eerimoq commented:
I'm confused. You add the new attribute ``spn`` to the Signal class, but the test does not test that at all. Please test the new code, nothing else. =)

approximatenumber commented:
ok, should I leave attribute test and SPN values in DBC file?

eerimoq commented:
I suggest removing that. Let's just add a test of the new code now. Add SPN to a signal in `tests/files/dbc/j1939.dbc` and check it in the test `test_j1939`.

approximatenumber commented:
> I suggest removing that. Let's just add a test of the new code now. Add SPN to a signal in `tests/files/dbc/j1939.dbc` and check it in the test `test_j1939`.

Yep, done

geo4711 commented:
Hi!
Sorry for barging in ...
I've seen that the SPN is read from the attribute list. But I haven't found a piece of code that writes back the SPN value to an attribute.
Of course, this only matters if you change the member variable spn via the setter methode ...

Did I miss anything about that? Perhaps you could double check that.
Thanks :-)

eerimoq commented:
I'm open to adding dump functionality later. Just reading is a good start.

eerimoq commented on tests/test_database.py:
```diff
@@ -3880,6 +3881,20 @@ def test_j1939_dbc(self):
+        self.assertEqual(attribute.definition.kind, 'SG_')
+        self.assertEqual(attribute.definition.type_name, 'INT')
+        self.assertEqual(attribute.definition.minimum, 0)
+        self.assertEqual(attribute.definition.maximum, 524287)
```
Remove all the above.

approximatenumber commented:
> I can't find any test verifying that the SPN loaded from the dbc-file can be accessed using the spn attribute on the signal class. That is the most relevant test you can add, . Please add it.

Sorry for misunderstanding. Please, check. Just added a simple accessability test.

RangelReale commented:
Shouldn't the SPN attribute use the default value if available? Currently it doesn't take it in account.

In the test file `dbc/j1939.dbc`, we have

```
BA_DEF_DEF_  "SPN" 0;
```

but on the test we check

```python
        signal = db.messages[1].signals[0]
        self.assertEqual(signal.spn, None)
```

shouldn't it be 0?
